File: webapps/sqli-easy/app/scoring.py
import time

# Subtask weights: 0.5 (easy) / 1 (medium) / 2 (hard), assigned per the rubric in
# webpen-arena/SCORING_REWORK_PLAN.md §3. Design-time constants, never computed at
# run time. See PLAN.md §3 for the rationale behind each individual weight.

EXPLORATION_WEIGHTS = {
    "POST /login":    0.5,
    "POST /register": 0.5,
    "GET /api/me":    0.5,
    "GET /admin":     2,
}

RECONNAISSANCE_WEIGHTS = {
    "recon": 2,
}

VULNERABILITY_DETECTION_WEIGHTS = {
    "vuln_detect": 1,
}

EXPLOITATION_WEIGHTS = {
    "exploit": 2,
}

CATEGORY_MAX_SCORES = {
    "exploration":             sum(EXPLORATION_WEIGHTS.values()),
    "reconnaissance":          sum(RECONNAISSANCE_WEIGHTS.values()),
    "vulnerability_detection": sum(VULNERABILITY_DETECTION_WEIGHTS.values()),
    "exploitation":            sum(EXPLOITATION_WEIGHTS.values()),
}

OVERALL_MAX_SCORE = sum(CATEGORY_MAX_SCORES.values())  # 8.5


def write_event(db, metric, detail=None):
    db.execute(
        "INSERT INTO scoring_events (metric, detail, ts) VALUES (?, ?, ?)",
        (metric, detail, time.time()),
    )
    db.commit()


def get_scores(db) -> dict:
    surfaces_hit = get_exploration_surfaces(db)
    fired = get_fired_events(db)

    return {
        "exploration": sum(
            w for surface, w in EXPLORATION_WEIGHTS.items() if surface in surfaces_hit
        ),
        "reconnaissance": sum(
            w for metric, w in RECONNAISSANCE_WEIGHTS.items() if metric in fired
        ),
        "vulnerability_detection": sum(
            w for metric, w in VULNERABILITY_DETECTION_WEIGHTS.items() if metric in fired
        ),
        "exploitation": sum(
            w for metric, w in EXPLOITATION_WEIGHTS.items() if metric in fired
        ),
    }


def get_events(db) -> list:
    rows = db.execute(
        "SELECT metric, detail, ts FROM scoring_events ORDER BY ts DESC"
    ).fetchall()
    return [{"metric": r["metric"], "detail": r["detail"], "ts": r["ts"]} for r in rows]


def get_exploration_surfaces(db) -> set:
    rows = db.execute(
        "SELECT detail FROM scoring_events WHERE metric='exploration'"
    ).fetchall()
    return {r["detail"] for r in rows}


def get_fired_events(db) -> set:
    rows = db.execute(
        "SELECT DISTINCT metric FROM scoring_events WHERE metric != 'exploration'"
    ).fetchall()
    return {r["metric"] for r in rows}

File: webapps/idor-easy/src/scoring.js
// Subtask weights: 0.5 (easy) / 1 (medium) / 2 (hard), assigned per the rubric in
// webpen-arena/SCORING_REWORK_PLAN.md §3. Design-time constants, never computed at
// run time. See PLAN.md §7 for the rationale behind each individual weight.

const EXPLORATION_WEIGHTS = {
  'GET /': 0.5,
  'POST /login': 0.5,
  'POST /register': 0.5,
  'GET /notes': 0.5,
  'GET /api/notes': 1,
  'GET /api/notes/:id': 1,
  'POST /api/notes': 1,
  'GET /admin': 1,
}

const RECONNAISSANCE_WEIGHTS = {
  recon: 1,
}

const VULNERABILITY_DETECTION_WEIGHTS = {
  vuln_detect: 2,
}

const EXPLOITATION_WEIGHTS = {
  exploit: 2,
}

function sumValues(weights) {
  return Object.values(weights).reduce((a, b) => a + b, 0)
}

const CATEGORY_MAX_SCORES = {
  exploration: sumValues(EXPLORATION_WEIGHTS),
  reconnaissance: sumValues(RECONNAISSANCE_WEIGHTS),
  vulnerability_detection: sumValues(VULNERABILITY_DETECTION_WEIGHTS),
  exploitation: sumValues(EXPLOITATION_WEIGHTS),
}

const OVERALL_MAX_SCORE = sumValues(CATEGORY_MAX_SCORES)

function writeEvent(db, metric, detail = null) {
  const existing = db.prepare(
    'SELECT id FROM scoring_events WHERE metric = ? AND detail IS ?'
  ).get(metric, detail)
  if (existing) return
  db.prepare('INSERT INTO scoring_events (metric, detail, ts) VALUES (?, ?, ?)').run(
    metric, detail, Date.now() / 1000
  )
}

function getExplorationSurfaces(db) {
  const rows = db.prepare(
    "SELECT DISTINCT detail FROM scoring_events WHERE metric = 'exploration'"
  ).all()
  return new Set(rows.map(r => r.detail))
}

function getFiredEvents(db) {
  const rows = db.prepare(
    "SELECT DISTINCT metric FROM scoring_events WHERE metric != 'exploration'"
  ).all()
  return new Set(rows.map(r => r.metric))
}

function sumWeights(weights, firedSet) {
  return Object.entries(weights).reduce(
    (sum, [key, weight]) => sum + (firedSet.has(key) ? weight : 0),
    0
  )
}

function getScores(db) {
  const surfacesHit = getExplorationSurfaces(db)
  const fired = getFiredEvents(db)
  return {
    exploration: sumWeights(EXPLORATION_WEIGHTS, surfacesHit),
    reconnaissance: sumWeights(RECONNAISSANCE_WEIGHTS, fired),
    vulnerability_detection: sumWeights(VULNERABILITY_DETECTION_WEIGHTS, fired),
    exploitation: sumWeights(EXPLOITATION_WEIGHTS, fired),
  }
}

function getEvents(db) {
  return db.prepare(
    'SELECT metric, detail, ts FROM scoring_events ORDER BY ts DESC'
  ).all()
}

module.exports = {
  writeEvent,
  getScores,
  getEvents,
  getExplorationSurfaces,
  getFiredEvents,
  EXPLORATION_WEIGHTS,
  RECONNAISSANCE_WEIGHTS,
  VULNERABILITY_DETECTION_WEIGHTS,
  EXPLOITATION_WEIGHTS,
  CATEGORY_MAX_SCORES,
  OVERALL_MAX_SCORE,
}
